#nanopore_size_analysis.R
#this file is meant to be used inside Rstudio
#
#this file takes a flat table of insert sizes for each read from a bam file, and plots distributions and computes statistics
#
#the insert sizes come from mod_mappings.bam files, that are generated by megalodon and then sorted.
#
#insert sizes can be computed with something like this, as an example:
#samtools view -@ 20 P6600_plas_mod_mappings.sorted.bam | awk '{print ($1,length($10),"P6600_plas_mod_mappings.sorted.bam")}' > P6600_plas_mod_mappings.lengths.txt

library(data.table)
library(dplyr)
library(ggplot2)
library(tidyverse)

#read in length files
files <- c(Sys.glob("/mnt/ix1/Projects/M083_20211228_nanopore_methylation_GI/blood_center_and_CLs/00_samples/*.lengths.txt"),Sys.glob("/mnt/ix1/Projects/M083_20211228_nanopore_methylation_GI/tissue_bank_CRC/00_samples/*.lengths.txt"))

#sample 50000 reads in each one
out <- data.table()
for (i in 1:length(files)) {
  print(i)
  file <- files[i]
  sample <- basename(file)
  sample <- gsub(pattern = "_mod_mappings.lengths.txt", replacement = "", x=sample)
  tmp <- fread(cmd=paste("shuf",file), nrows = 50000) %>% mutate(sample=sample) 
  tmp$V3 <- NULL
  colnames(tmp) <- c("readname","length","sample")
  out <- bind_rows(out,tmp)
}

out2 <- out %>% filter(length < 600)
out2$sample <- factor(out2$sample, unique(out2$sample))

out2 <- out2 %>% mutate(type=ifelse(grepl("plas",sample),"Healthy","Cancer"))
out2$label <- paste(out2$sample, out2$type, sep = "\n")

label <- out2$label
names(label) <- out2$sample

#plot sizes
ggplot(out2 %>% filter(sample != "P6604_plas"), aes(x=length, fill=sample)) +
  facet_wrap(~sample, nrow=6, labeller = as_labeller(label)) +
  geom_density() +
  #geom_histogram() +
  xlab("Insert size (bp)") +
  ylab("Density") +
  theme_classic() +
  scale_y_continuous(breaks=c(0,5,10)) +
  scale_x_continuous(breaks=c(50,100,300), trans="log10") +
  geom_vline(xintercept = c(167, 334), linetype="dotted") +
  theme(legend.position="none",
        panel.margin.y = unit(0, "lines"),
        panel.margin.x = unit(0, "lines"),
        axis.text.x = element_text(size=20),
        axis.text.y = element_text(size=20),
        axis.title.x = element_text(size=40),
        axis.title.y = element_text(size=40),
        strip.text = element_text(size=20))

ggsave("tb_insert_size_densities_3.png", dpi=300, height=12, width=12, units="in")


#cherry pick to look at for main figure
labels =c(
  "P6600_plas" = "Healthy",
  "P2592_3329A" = "Cancer"
)

ggplot(out2 %>% filter(sample %in% c("P6600_plas", "P2592_3329A")), aes(x=length, fill=sample)) +
  facet_wrap(~sample, nrow=2, labeller =as_labeller(labels)) +
  geom_density() +
  xlab("Insert size (bp)") +
  ylab("Density") +
  theme_classic() +
  scale_y_continuous(breaks=c(0,5,10)) +
  scale_x_continuous(breaks=c(50,100,300), trans="log10") +
  geom_vline(xintercept = c(167, 334), linetype="dotted") +
  theme(legend.position="none",
        panel.margin.y = unit(0, "lines"),
        panel.margin.x = unit(0, "lines"),
        axis.text.x = element_text(size=20),
        axis.text.y = element_text(size=20),
        axis.title.x = element_text(size=40),
        axis.title.y = element_text(size=40),
        strip.text = element_text(size=30))

ggsave("tb_insert_size_densities3.png", dpi=300, height=12, width=6, units="in")

#compute mono-nucleosome and di-nucleosome ratios
out2 <- out2 %>% mutate(type=ifelse(grepl("plas",sample),"Blood Center","Tissue Bank"))
out2_nucleosomes <- out2 %>% group_by(type,sample) %>% summarize(mono=sum(length<=250.5)/n(), di=sum(length>250.5)/n(), mono_cv=sd(length[length<=250.5])/mean(length[length<=250.5]), di_cv=sd(length[length>250.5])/mean(length[length>250.5]))
out2_nucleosomes$ratio <- out2_nucleosomes$mono/out2_nucleosomes$di

out2_nucleosomes_melt <- melt(out2_nucleosomes, measure.vars=c("ratio"))

ggplot(out2_nucleosomes, aes(x=type,y=ratio, fill=type)) +
  geom_boxplot() +
  theme_classic(40) +
  ylab("Mononucleosome to\nDinucleosome ratio") +
  #scale_x_discrete(labels=c("Blood\nCenter","Tissue\nBank")) +
  scale_x_discrete(labels=c("Healthy","Cancer")) +
  theme(axis.text.x = element_text(size=40),
        axis.text.y = element_text(size=20),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=40),
        legend.position="none")

ggsave("tb_sbc_plasma_nucleosome_ratio4.png", dpi=300, height=12, width=7.5, units="in")

